File: order_book.py

class BookEntry:
  def __init__(self, price, size, volume):
    self.price = price
    self.size = size
    self.volume = volume
    self.book_sum = 0

  def update(self, order):
    self.size += order.size
    self.volume += order.bitcoins_desired

  def __str__(self):
    return " ".join([str(self.price),
      str(self.size),
      str(self.volume),
      str(self.book_sum)])


class OrderBook:
  def __init__(self):
    self.bids = []
    self.offers = []
    self.eps = 0.00000001

  def place_sell_order(self, order):
    self.offers.append(order)

  def place_buy_order(self, order):
    self.bids.append(order)

  def aggregate_sell_orders(self):
    sorted_list = sorted(self.offers, key=lambda order: order.price)
    return sorted_list

  def accumulate_book(self):
    sorted_list = self.aggregate_sell_orders()
    book = []

    if len(sorted_list) == 0:
      return book

    i = 0
    order = sorted_list[i]
    be = BookEntry(order.price, order.size, order.bitcoins_desired)
    while True:
      if (i+1) >= len(sorted_list):
        book.append(be)
        break
      else:
        i += 1
        next_order = sorted_list[i]
        if next_order.price - order.price < self.eps:
          be.update(next_order)
        else:
          book.append(be)
          order = next_order
          be = BookEntry(order.price, order.size, order.bitcoins_desired)

    book[0].book_sum = sorted_list[0].bitcoins_desired
    for i in xrange(1, len(book)):
      book[i].book_sum = book[i-1].book_sum + book[i].volume

    return book

File: OrderBook.js
var EventEmitter = require('events').EventEmitter;
var MXOrderScraper = require('./MXOrderScraper.js');

function BookEntry(price, size) {
  this.price = price;
  this.size = size;
  this.bookSum = 0;
}

BookEntry.prototype.update = function(other) {
  this.size += other.size;
}

function compareOrdersAsc(a,b) {
  if(a.price < b.price) return -1;
  if(a.price > b.price) return 1;
  return 0;
}

function aggregateBuys(book) {
  var len = book.length;
  book[len-1].bookSum = book[len-1].size
  for(var i = len - 2; i >= 0; i--) {
    book[i].bookSum = book[i+1].bookSum + book[i].size;
  }
  return book;
}

function aggregateSells(book) {
  book[0].bookSum = book[0].size;
  for(var i = 1; i < book.length; i++) {
    book[i].bookSum = book[i-1].bookSum + book[i].size;
  }
  return book;
}

function OrderBook(source) {
  this.eps = 0.001;

  var self = this;
  source.on("buys", function(orders) {
    var buyBook = aggregateBuys(self.aggregateBook(orders));
    self.emit("buys", buyBook);
  });
  source.on("sells", function(orders) {
    var sellBook = aggregateSells(self.aggregateBook(orders));
    self.emit("sells", sellBook);
  });
}
OrderBook.prototype = Object.create(EventEmitter.prototype);


OrderBook.prototype.aggregateBook = function(orders) {
  orders.sort(compareOrdersAsc);
  var book = [];
  if(orders.length == 0) return book;

  var i = 0;
  var order = orders[i];
  var be = new BookEntry(order.price, order.size);
  while(true) {
    if((i+1) >= orders.length) {
      book.push(be);
      break;
    } else {
      i++;
      next = orders[i];
      if(next.price - order.price < this.eps) {
        be.update(next);
      } else {
        book.push(be);
        order = next;
        be = new BookEntry(order.price, order.size);
      }
    }
  }
  return book;
}


module.exports = OrderBook;
